File: app/src/main/java/fr/free/nrw/commons/notification/NotificationClient.java
package fr.free.nrw.commons.notification;

import fr.free.nrw.commons.notification.models.Notification;
import org.wikipedia.csrf.CsrfTokenClient;
import org.wikipedia.dataclient.Service;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.Single;
import org.wikipedia.dataclient.mwapi.MwQueryResponse;

import static fr.free.nrw.commons.di.NetworkingModule.NAMED_COMMONS_CSRF;

@Singleton
public class NotificationClient {

    private final NotificationInterface service;
    private final CsrfTokenClient csrfTokenClient;

    @Inject
    public NotificationClient(
        final NotificationInterface service,
        @Named(NAMED_COMMONS_CSRF) final CsrfTokenClient csrfTokenClient) {
        this.service = service;
        this.csrfTokenClient = csrfTokenClient;
    }

    public Single<List<Notification>> getNotifications(final boolean archived) {
        return service.getAllNotifications("wikidatawiki|commonswiki|enwiki", archived ? "read" : "!read", null)
                .map(mwQueryResponse -> mwQueryResponse.query().notifications().list())
                .flatMap(Observable::fromIterable)
                .map(Notification::from)
                .toList();
    }

    public Observable<Boolean> markNotificationAsRead(final String notificationId) {
        try {
            return service.markRead(csrfTokenClient.getTokenBlocking(), notificationId, "")
                    .map(MwQueryResponse::success);
        } catch (final Throwable throwable) {
            return Observable.just(false);
        }
    }
}

File: app/src/main/java/fr/free/nrw/commons/notification/NotificationClient.kt
package fr.free.nrw.commons.notification

import fr.free.nrw.commons.di.NetworkingModule
import fr.free.nrw.commons.notification.models.Notification
import fr.free.nrw.commons.notification.models.NotificationType
import io.reactivex.Observable
import io.reactivex.Single
import fr.free.nrw.commons.auth.csrf.CsrfTokenClient
import fr.free.nrw.commons.auth.csrf.InvalidLoginTokenException
import fr.free.nrw.commons.wikidata.mwapi.MwQueryResponse
import fr.free.nrw.commons.utils.DateUtil
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton
import fr.free.nrw.commons.wikidata.model.notifications.Notification as WikimediaNotification

@Singleton
class NotificationClient @Inject constructor(
    @param:Named(NetworkingModule.NAMED_COMMONS_CSRF) private val csrfTokenClient: CsrfTokenClient,
    private val service: NotificationInterface
) {
    fun getNotifications(archived: Boolean): Single<List<Notification>> =
        service.getAllNotifications(
            wikiList = "wikidatawiki|commonswiki|enwiki",
            filter = if (archived) "read" else "!read",
            continueStr = null
        ).map {
            it.query()?.notifications()?.list() ?: emptyList()
        }.flatMap {
            Observable.fromIterable(it)
        }.map {
            it.toCommonsNotification()
        }.toList()

    fun markNotificationAsRead(notificationId: String?): Observable<Boolean> {
        return try {
            service.markRead(
                token = csrfTokenClient.getTokenBlocking(),
                readList = notificationId,
                unreadList = ""
            ).map(MwQueryResponse::success)
        } catch (throwable: Throwable) {
            if (throwable is InvalidLoginTokenException) {
                Observable.error(throwable)
            } else {
                Observable.just(false)
            }
        }
    }

    private fun WikimediaNotification.toCommonsNotification() = Notification(
        notificationType = NotificationType.UNKNOWN,
        notificationText = contents?.compactHeader ?: "",
        date = DateUtil.getMonthOnlyDateString(timestamp),
        link = contents?.links?.primary?.url ?: "",
        iconUrl = "",
        notificationId = id().toString()
    )
}
